### PR TITLE
[IN:12672] Automate SpiraApp deployment with version auto-increment, interactive mode selection (build/enable/disable), project-lvl mng via env variables, and Playwright-driven browser automation for the complete upload and activation workflow

### PR DESCRIPTION
**Pull request by destapickering-9504**
Spira Token: [[IN:12672](https://inflectra.spiraservice.net/75/Incident/12672.aspx?referrerProjectList=true)] 
## Summary

This PR adds a new developer automation script (`bundle-automation.js`) that streamlines the SpiraApp development workflow by automating building, uploading, and deployment to Spira instances. The script is designed to help developers quickly iterate and test their SpiraApps.

### 🆕 New File: `bundle-automation.js`
A comprehensive automation script (576 lines) that handles the complete SpiraApp deployment workflow:

**Key Features:**
- ✅ Automated version bumping (e.g., 1.0 → 1.1)
- ✅ Build integration with existing `index.js`
- ✅ Headless browser automation using Playwright
- ✅ Automatic login and Developer Mode enablement
- ✅ SpiraApp upload and system-wide activation
- ✅ Project-specific enable/disable functionality
- ✅ Three operation modes: full automation, enable-only, disable-only

**Code Quality:**
- Well-structured with helper functions (`normalizeBaseUrl`, `readManifest`, `withBrowser`)
- Comprehensive error handling and validation
- Detailed JSDoc comments
- Modular design with reusable functions

### 🧪 New File: `test/bundle-automation.test.js`
Complete test suite (253 lines) covering:
- CLI argument parsing
- Output folder derivation
- Environment variable loading and validation
- Version bumping logic
- Edge cases and error conditions

**Test Results:** ✅ 21 tests, all passing

### 📦 Dependencies Added
- `playwright` - Browser automation
- `dotenv` - Environment variable management
- `js-yaml` - YAML parsing
- `fast-check` - Property-based testing (dev dependency)

### 📝 Documentation Updates (`readme.md`)
- **Restructured for clarity** - Separated manual builds from developer automation
- **Added "Developer Automation" section** - Complete guide for using `bundle-automation.js`
- **Environment configuration guide** - Clear instructions for `.env` setup
- **Usage examples** - Practical examples for all three operation modes
- **Corrected build commands** - Fixed syntax for `npm run build`

### 🔧 Configuration Files
- `.env.example` - Template for required environment variables
- `.kiro/specs/bundle-automation/` - Spec files documenting requirements, design, and tasks

### Manual Build (Existing Workflow)
```bash
npm run build -- --input=/path/to/SpiraApp --output=/path/to/output

### CODE REVIEW

File: .kiro/specs/bundle-automation/.config.kiro
Comment (simonhbor) [inline .kiro/specs/bundle-automation/.config.kiro:1]: As this is a public repo, I think we should remove all kiro specs

File: readme.md [hunk 1]
```diff
@@ -3,18 +3,79 @@ This repo takes a SpiraApp manifest.yaml describing a SpiraApp and its associate
...
+- `SPIRA_PASSWORD` - Your Spira password (required)
+- `SPIRA_ENABLE_PROJECT_IDS` - Comma-separated project IDs to enable the SpiraApp for (optional)
+- `SPIRA_DISABLE_PROJECT_IDS` - Comma-separated project IDs to disable the SpiraApp for (optional)
+
+### Usage
+
+Run the automation script:
+
+```bash
+npm run automate -- --input="C:\MySpiraApp"
```
Comment (simonhbor) [inline readme.md]: should it really be `-- --input`?

File: readme.md [hunk 2]
```diff
@@ -3,18 +3,79 @@ This repo takes a SpiraApp manifest.yaml describing a SpiraApp and its associate
...
+```
+
+The script will prompt you to choose a mode:
+
+1. **Build, upload and enable** - Builds the package, uploads it to Spira, enables it system-wide, and enables it for projects specified in `SPIRA_ENABLE_PROJECT_IDS`
+2. **Enable only** - Enables an already-uploaded SpiraApp for projects specified in `SPIRA_ENABLE_PROJECT_IDS` (no build or upload)
+3. **Disable only** - Disables the SpiraApp for projects specified in `SPIRA_DISABLE_PROJECT_IDS`
```
Comment (simonhbor) [inline readme.md]: I get the benefit of all of these options, But what about just "build and upload"? That to me feels like the option I would use most of the time and should be #1. I first turn it on for a product, then after each code change I will want to just replace it on the admin page.

Also can you confirm it runs headless? Or is there a way to set that up?

File: readme.md [hunk 3]
```diff
@@ -3,18 +3,79 @@ This repo takes a SpiraApp manifest.yaml describing a SpiraApp and its associate
...
+
+You can skip the prompt by using command-line flags:
+
+```bash
+# Enable only (no build/upload)
+npm run automate -- --input="C:\MySpiraApp" --enable
+
+# Disable only
+npm run automate -- --input="C:\MySpiraApp" --disable
```
Comment (simonhbor) [inline readme.md]: Is there a command line flag for building and uploading? Maybe just called --upload?

File: readme.md [hunk 4]
```diff
@@ -3,18 +3,79 @@ This repo takes a SpiraApp manifest.yaml describing a SpiraApp and its associate
...
+```
+
+### What the Automation Does
+
+- Automatically increments the patch version in manifest.yaml (e.g., 1.0 → 1.1)
+- Builds the .spiraapp package file
+- Logs into your Spira instance using Playwright browser automation
+- Enables Developer Mode in Spira (if not already enabled)
+- Uploads the .spiraapp file to the SpiraApps administration page
+- Activates the SpiraApp system-wide
```
Comment (simonhbor) [inline readme.md]: Can we clarify that if it is already active it will not accidentally deactivate it?

File: bundle-automation.js [hunk 5]
```diff
@@ -0,0 +1,628 @@
...
+ * Ensure Developer Mode is enabled in Spira General Settings.
+ *
+ * Navigates to the General Settings admin page, checks the Developer Mode
+ * checkbox/toggle, enables it if not already on, and saves. If already
+ * enabled, proceeds without modifying the setting.
+ *
+ * @param {import('playwright').Page} page
+ * @param {{ baseUrl: string }} config
+ */
+async function enableDeveloperMode(page, config) {
```
Comment (simonhbor) [inline bundle-automation.js:406]: Can we skip this if a user wants? This is great for the first time, and not sure where this is called, but I think it would be great if we can make the uploaded of a new version as quick and painless as possible during development

File: readme.md [hunk 6]
```diff
@@ -3,18 +3,79 @@ This repo takes a SpiraApp manifest.yaml describing a SpiraApp and its associate
...
+# Enable only (no build/upload)
+npm run automate -- --input="C:\MySpiraApp" --enable
+
+# Disable only
+npm run automate -- --input="C:\MySpiraApp" --disable
+```
+
+### What the Automation Does
+
+- Automatically increments the patch version in manifest.yaml (e.g., 1.0 → 1.1)
```
Comment (simonhbor) [inline readme.md]: Can we disable this? NOTE that incrementing of version number is not required for a new SpiraApp to take effect when it is a .spiraapp file. During development we do not always want this to happen - usually we do not. 

Maybe instead it could be in the .env file? A bool to increment version or not? Or an option in the CLI / as an extra cli-flag?

**Review by simonhbor — CHANGES_REQUESTED** [review]
This is in great shape. I have some suggestions for improving the docs and mainly the underlying feature set to make this more usable.

Also we should not have our kiro work in the public and this a public repo, so I think we should delete those files. For private repos it is fine. Suggest adding to .gitignore unfortunately.

File: readme.md (hunk 4 above)
Comment (destapickering-9504) [reply readme.md]: Clarify, as in make note of it in writing? In code, if it's activated, it can't be deactivated by running enable again. Likewise with disable. The code looks for a specific flag which only appears when it's enabled or disabled.

File: readme.md (hunk 2 above)
Comment (destapickering-9504) [reply readme.md]: If project id's are not provided, we will still build and upload and enable it for system wide consumption. If there are too many options, I can remove them. I was trying to provide ease of use. If it's not easy, happy to remove it. 

It currently runs headless, but I can change that to be a flag if you think people would like to see the click ops. Let me know your opinion. IMO: having it run headless makes developer sense.

File: .kiro/specs/bundle-automation/.config.kiro
Comment (destapickering-9504) [reply .kiro/specs/bundle-automation/.config.kiro:1]: Makes sense. I agree.

File: readme.md (hunk 1 above)
Comment (destapickering-9504) [reply readme.md]: Will remove... It's a dated way I was handling it earlier.

File: readme.md (hunk 3 above)
Comment (destapickering-9504) [reply readme.md]: Would upload need to check that developer mode is enabled as well? Would the ability to not need to add a project id not be enough for an --upload.

File: bundle-automation.js (hunk 5 above)
Comment (destapickering-9504) [reply bundle-automation.js:406]: Yes, I will add a flag for this.

File: readme.md (hunk 6 above)
Comment (destapickering-9504) [reply readme.md]: I'm not sure if we can disable this... The page was having a hard time discovering if the app was successfully upload when a repeat package was uploaded. But I can take a deeper look into this if it's a no go.

**Comment by destapickering-9504** [comment]
I have updated your PR review. I have addressed all your concerns, please let me know if there is anything else you are worried about.

File: .env.example [hunk 7]
```diff
@@ -5,3 +5,9 @@ SPIRA_PASSWORD=your_password
 SPIRA_ENABLE_PROJECT_IDS=
 # Optional: comma-separated project IDs to disable the SpiraApp for (e.g. 1,2,3)
 SPIRA_DISABLE_PROJECT_IDS=
+# Optional: set to true to run the browser in headless mode; default is headed (non-headless)
+PLAYWRIGHT_HEADLESS=
```
Comment (simonhbor) [inline .env.example]: Sorry for any confusion here, but I think that headless should be the default. Not even sure if we need it to be headed ever. I only asked earlier to know what it did. Headless is better. But a setting is no bad thing, I just think we should change default behavior

File: .env.example
Comment (simonhbor) [inline .env.example:1]: Given this is an example file can you please add example values like you show for the IDs? Or put the allowed values in comments. Basically should the 3 final values all take true or false?

File: .env.example [hunk 8]
```diff
@@ -5,3 +5,9 @@ SPIRA_PASSWORD=your_password
...
+# Optional: set to true to enable Developer Mode in Spira General Settings before uploading; default is skipped
```
Comment (simonhbor) [inline .env.example]: default is skipped which also means false? Maybe add " (false)" at end

File: .env.example [hunk 9]
```diff
@@ -5,3 +5,9 @@ SPIRA_PASSWORD=your_password
...
+SPIRA_ENABLE_DEV_MODE=
+# Optional: set to true to auto-increment the patch version in manifest.yaml before each build; default is to skip the version bump
```
Comment (simonhbor) [inline .env.example]: default is skipped which also means false? Maybe add " (false)" at end

File: readme.md [hunk 10]
```diff
@@ -1,81 +1,172 @@
...
+
+This script is designed for developers who want to quickly iterate and test SpiraApps without manually uploading through the UI each time.
 
 ### Environment Configuration
 
-Create a `.env` file in the repo root with the following variables:
+Copy `.env.example` to `.env` and fill in your values:
 
 ```
+# Required
```
Comment (simonhbor) [inline readme.md]: I say delete this section and just rely on the table which is more comprehensive

File: readme.md [hunk 11]
```diff
@@ -1,81 +1,172 @@
...
-- `SPIRA_DISABLE_PROJECT_IDS` - Comma-separated project IDs to disable the SpiraApp for (optional)
+| Variable | Required | Description |
+|---|---|---|
+| `SPIRA_BASE_URL` | Yes | Base URL of your Spira instance |
+| `SPIRA_USERNAME` | Yes | Your Spira username |
+| `SPIRA_PASSWORD` | Yes | Your Spira password |
+| `SPIRA_ENABLE_PROJECT_IDS` | No | Comma-separated project IDs to enable the SpiraApp for |
+| `SPIRA_DISABLE_PROJECT_IDS` | No | Comma-separated project IDs to disable the SpiraApp for |
+| `PLAYWRIGHT_HEADLESS` | No | Set to `true` to run the browser in headless mode. Default is headed (visible browser). |
+| `SPIRA_ENABLE_DEV_MODE` | No | Set to `true` to enable Developer Mode in Spira General Settings before uploading. Default is skipped — only needed the first time or if Developer Mode has been turned off. |
```
Comment (simonhbor) [inline readme.md]: For this line and below please clarify if the default value is true or false as I think it is a bit unclear - eg add (false) to end of the lines if that is correct

File: readme.md [hunk 12]
```diff
@@ -1,81 +1,172 @@
...
+
+Logs into Spira and enables the SpiraApp for the projects listed in `SPIRA_ENABLE_PROJECT_IDS`. No build or upload is performed.
+
+**Disable only (`--disable`)**
+
+Logs into Spira and disables the SpiraApp for the projects listed in `SPIRA_DISABLE_PROJECT_IDS`. No build or upload is performed.
+
+### Important Notes
+
+**Activation is idempotent — it does not toggle**
```
Comment (simonhbor) [inline readme.md]: I honestly did not know the word `idempotent`. I probably should have and do now. Given this is a mini heading maybe we just says "Enable/Disable" as this aligns with verbs we already use?

File: readme.md [hunk 13]
```diff
@@ -1,81 +1,172 @@
...
-- Builds the .spiraapp package file
-- Logs into your Spira instance using Playwright browser automation
-- Enables Developer Mode in Spira (if not already enabled)
-- Uploads the .spiraapp file to the SpiraApps administration page
-- Activates the SpiraApp system-wide
-- Enables or disables the SpiraApp for specific projects based on environment variables
+**Headless mode**
 
-**Note**: Inflectra will never ask you for your spiraapp file, only ever the source code that is used to generate it 
+By default the browser runs in headed mode (you can see it). Set `PLAYWRIGHT_HEADLESS=true` to run silently. In headless mode, errors are still reported to the terminal with clear, actionable messages.
```
Comment (simonhbor) [inline readme.md:158]: As flagged above I think headless should be the default. Apologies if things were not clear about that.

**Review by simonhbor — CHANGES_REQUESTED** [review]
Some minor suggestions to the readme. And I think headless should be the default mode. I did not mean to imply otherwise yesterday.

Other changes all make sense and I think this is looking great and will help people out a lot.

File: .env.example
Comment (destapickering-9504) [reply .env.example:1]: Yes, I'll put the defaults down.

File: .env.example (hunk 7 above)
Comment (destapickering-9504) [reply .env.example]: I'll set it to a default, but I think having the option may be a good thing.

File: .env.example (hunk 8 above)
Comment (destapickering-9504) [reply .env.example]: resolved

File: .env.example (hunk 9 above)
Comment (destapickering-9504) [reply .env.example]: resolved

File: readme.md (hunk 10 above)
Comment (destapickering-9504) [reply readme.md]: resolved

File: readme.md (hunk 11 above)
Comment (destapickering-9504) [reply readme.md]: resolved

File: readme.md (hunk 12 above)
Comment (destapickering-9504) [reply readme.md]: resolved

File: readme.md (hunk 13 above)
Comment (destapickering-9504) [reply readme.md:158]: resolved

**Review by simonhbor — APPROVED** [review]

**Comment by simonhbor** [comment]
@Pecant-Pie  please assign back to me when you think the QA is done and it is ready for final review

**Comment by Pecant-Pie** [comment]
Verified that the issue with building SpiraApps with different values for 'name' and 'caption' has been fixed. There is also now a clear error message if the url in your .env file does not point to an existing web page. These were the only issues I noted before, so I think this is ready for final review. Handing this back to @simonhbor.

**Review by Pecant-Pie — APPROVED** [review]
